quint/src/effects/inferrer.ts

```diff
@@ -192,14 +193,17 @@ export class EffectInferrer implements IRVisitor {
         const substitution = arrowEffect.chain(effect => unify(signature, effect))
 
         const resultEffectWithSubs = substitution
-          .chain(s => compose(s, this.substitutions))
+          .chain(s => compose(this.substitutions, s))
           .chain(s => {
             this.substitutions = s
 
-            this.effects.forEach((effect, id) => {
-              const r = applySubstitutionToScheme(s, effect)
+            paramsResult.map(effects => zip(effects, expr.args.map(a => a.id)).forEach(([effect, id]) => {
+              if (!effect || !id) {
+                throw new Error('Invalid arrays')
+              }
+              const r = applySubstitution(s, effect).map(toScheme)
```

Comment:
I think a line break for the body of the mapped function can help indicate the scope of the mapped function here:

```suggestion
            paramsResult.map(effects => 
              zip(effects, expr.args.map(a => a.id)).forEach(([effect, id]) => {
                if (!effect || !id) {
                  throw new Error('Invalid arrays')
                }
                const r = applySubstitution(s, effect).map(toScheme)
```

otherwise, i need to count parens to figure out if we have of the `map(zip).forEach` or `map(zip.forEach)`.